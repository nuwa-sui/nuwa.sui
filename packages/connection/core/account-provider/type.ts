import type { SuiAddress } from "@nuwa.sui/types"
import type { Transaction } from "@mysten/sui/transactions"
import type { SuiChain, SuiFeatures, StandardFeatures, WalletAccount } from '@mysten/wallet-standard'

export type AllFeaturesString = keyof SuiFeatures | keyof StandardFeatures

export interface SuiAccount {
    address: SuiAddress
    /** find path: domain > name > address */
    label: string
    chains: SuiChain[]
    name?: string
    icon_url?: string
    publicKey?: Uint8Array | string | any
    features?: AllFeaturesString[]
}

export interface AccountProvider {
    get accounts(): WalletAccount[]
    readonly account: WalletAccount | undefined
    readonly chain: SuiChain | undefined

    readonly meta: {
        /** wallet name */
        readonly name: string | undefined
        /** wallet icon url */
        readonly icon_url: string | undefined
        readonly supported_chains: SuiChain[]
        readonly supported_features: AllFeaturesString[]
    }

    setChain(chain: SuiChain): void
    setAccount(arg: SuiAddress | WalletAccount | ((accounts: WalletAccount[]) => WalletAccount)): void
    onDisconnect(callback: () => void): void
    onAccountChange(callback: (accounts: WalletAccount[]) => void): void

    disconnect(): Promise<void>
    signTransaction(tx: Transaction): Promise<Uint8Array>
    signPersonalMessage(message: Uint8Array | string): Promise<Uint8Array>
    reportTransactionEffects(rawEffect: Uint8Array): Promise<void>
}