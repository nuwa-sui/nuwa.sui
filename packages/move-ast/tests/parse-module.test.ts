import type { Storage } from 'unstorage'
import { MoveParser2024, tokenizeMove } from '@'
import { createStorage } from 'unstorage'
import githubDriver from 'unstorage/drivers/github'
import { expect } from 'vitest'

it('should module be parsed', ({ parser }) => {
    parser.input = tokenizeMove(
        `
        // some comments
        
        module TestPackage::TestModule {
        
            #[attr]
            public struct TestResource {
                value: u8
            }
        }
        `,
    )
    expect(parser.parseModule()).toMatchObject({})

    parser.input = tokenizeMove(
        `
        module test::Abc;
        
        #[attr]
        public struct TestResource {
            value: u8
        }
        `,
    )
    expect(parser.parseModule()).toMatchObject({})
})

it('should fix error reference', ({ parser }) => {
    parser.input = tokenizeMove(
        `
        module TestPackage::TestModule;
        
        
        public fun abc(arg: TestA): TestB {
        }
        
        public struct TestB {
            value: TestA
        }
        
        public struct TestA {
            value: SUI
        }
        
        use sui::sui::SUI;
        `,
    )

    const module = parser.parseModule()

    expect(module).toMatchSnapshot()
})

const storageSuiFramework = createStorage({
    driver: githubDriver({
        repo: 'MystenLabs/sui',
        dir: 'crates/sui-framework/packages/move-stdlib/sources',
        branch: 'testnet',
    }),
})

const storageSuiStdlib = createStorage({
    driver: githubDriver({
        repo: 'MystenLabs/sui',
        dir: 'crates/sui-framework/packages/sui-framework/sources',
        branch: 'testnet',
    }),
})

async function toStorageKeyPair(storage: Storage<string>, name: string): Promise<[Storage<string>, string, string][]> {
    const rst: [Storage<string>, string, string][] = []
    for (const key of await storage.keys()) {
        rst.push([storage, key, name])
    }
    return rst
}

it.concurrent.each<[Storage<string>, string, string]>([
    ...(await toStorageKeyPair(storageSuiFramework, 'sui-framework')),
    ...(await toStorageKeyPair(storageSuiStdlib, 'sui-stdlib')),
])('should parse std-lib success', async (storage, key, name) => {
    consola.info('fetch: ', name, ' - ', key)
    const sourceCode = (await storage.get(key)) as string
    const parser = new MoveParser2024(tokenizeMove(sourceCode))
    const module = parser.parseModule()
    expect(module).toMatchObject({})
    expect(parser.errors.length, JSON.stringify(parser.errors)).toBe(0)
    consola.success('parse success: ', name, ' - ', key)
})
