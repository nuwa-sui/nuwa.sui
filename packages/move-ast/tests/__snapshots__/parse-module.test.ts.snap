// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should fix error reference 1`] = `
{
  "functions": {
    "abc": {
      "generics": [],
      "modifiers": [
        "public",
      ],
      "params": {
        "arg": {
          "generics": [],
          "ref": "value",
          "resource": {
            "abilities": [],
            "fields": {
              "value": {
                "generics": [],
                "ref": "value",
                "resource": {
                  "target": "sui::sui::SUI",
                  "type": "imported",
                },
              },
            },
            "generics": [],
            "public": true,
            "target": "TestPackage::TestModule::TestA",
            "type": "struct",
            "unique": false,
          },
        },
      },
      "returns": [
        {
          "generics": [],
          "ref": "value",
          "resource": {
            "abilities": [],
            "fields": {
              "value": {
                "generics": [],
                "ref": "value",
                "resource": {
                  "abilities": [],
                  "fields": {
                    "value": {
                      "generics": [],
                      "ref": "value",
                      "resource": {
                        "target": "sui::sui::SUI",
                        "type": "imported",
                      },
                    },
                  },
                  "generics": [],
                  "public": true,
                  "target": "TestPackage::TestModule::TestA",
                  "type": "struct",
                  "unique": false,
                },
              },
            },
            "generics": [],
            "public": true,
            "target": "TestPackage::TestModule::TestB",
            "type": "struct",
            "unique": false,
          },
        },
      ],
      "target": "TestPackage::TestModule::abc",
    },
  },
  "moduleName": "TestModule",
  "packageName": "TestPackage",
  "resources": {
    "TestA": {
      "abilities": [],
      "fields": {
        "value": {
          "generics": [],
          "ref": "value",
          "resource": {
            "target": "sui::sui::SUI",
            "type": "imported",
          },
        },
      },
      "generics": [],
      "public": true,
      "target": "TestPackage::TestModule::TestA",
      "type": "struct",
      "unique": false,
    },
    "TestB": {
      "abilities": [],
      "fields": {
        "value": {
          "generics": [],
          "ref": "value",
          "resource": {
            "abilities": [],
            "fields": {
              "value": {
                "generics": [],
                "ref": "value",
                "resource": {
                  "target": "sui::sui::SUI",
                  "type": "imported",
                },
              },
            },
            "generics": [],
            "public": true,
            "target": "TestPackage::TestModule::TestA",
            "type": "struct",
            "unique": false,
          },
        },
      },
      "generics": [],
      "public": true,
      "target": "TestPackage::TestModule::TestB",
      "type": "struct",
      "unique": false,
    },
  },
  "test": false,
}
`;
